File: backend-python/app/schemas/plugins.py
"""Plugin inventory REST API schemas (Epic 08 Phase 6).

Responses expose bounded inventory fields only — never manifest ``metadata``,
filesystem paths, entrypoint module paths, or template bodies.
"""

from __future__ import annotations

from pydantic import BaseModel, Field

from app.ai.plugins.models import PluginContributionKind, PluginStatus

__all__ = [
    "PluginDependencyResponse",
    "PluginInventoryDetail",
    "PluginInventoryItem",
    "PluginInventoryListResponse",
    "PluginLoadFailureResponse",
]


class PluginDependencyResponse(BaseModel):
    """Informational manifest dependency entry (detail endpoint only)."""

    plugin_id: str
    version: str | None = None


class PluginLoadFailureResponse(BaseModel):
    """Safe plugin load failure subset for REST responses."""

    code: str
    message: str
    expected_api_versions: list[str] | None = None
    manifest_api_version: str | None = None


class PluginInventoryItem(BaseModel):
    """One plugin record in the inventory list."""

    plugin_id: str | None = None
    name: str | None = None
    version: str | None = None
    api_version: str | None = None
    status: PluginStatus
    contributions: list[PluginContributionKind] = Field(default_factory=list)
    load_duration_ms: float = 0.0
    author: str | None = None
    homepage: str | None = None
    repository: str | None = None
    documentation: str | None = None
    license: str | None = None
    failure: PluginLoadFailureResponse | None = None


class PluginInventoryDetail(PluginInventoryItem):
    """Detail view for one resolvable plugin record."""

    dependencies: list[PluginDependencyResponse] = Field(default_factory=list)


class PluginInventoryListResponse(BaseModel):
    """Inventory list returned by ``GET /api/plugins``."""

    plugins: list[PluginInventoryItem] = Field(default_factory=list)

File: frontend/src/types/plugins.ts
export type PluginContributionKind = 'tool' | 'prompt' | 'workflow_node' | 'mcp_server'

export type PluginStatus = 'loaded' | 'failed'

export interface PluginLoadFailure {
  code: string
  message: string
  expected_api_versions?: string[] | null
  manifest_api_version?: string | null
}

export interface PluginInventoryItem {
  plugin_id: string | null
  name: string | null
  version: string | null
  api_version: string | null
  status: PluginStatus
  contributions: PluginContributionKind[]
  load_duration_ms: number
  author?: string | null
  homepage?: string | null
  repository?: string | null
  documentation?: string | null
  license?: string | null
  failure?: PluginLoadFailure | null
}

export interface PluginInventoryListResponse {
  plugins: PluginInventoryItem[]
}

export function displayPluginId(pluginId: string | null): string {
  return pluginId ?? 'Unknown plugin'
}

export function displayPluginName(item: PluginInventoryItem): string {
  return item.name ?? displayPluginId(item.plugin_id)
}

export function formatContributionKind(kind: PluginContributionKind): string {
  switch (kind) {
    case 'tool':
      return 'Tool'
    case 'prompt':
      return 'Prompt'
    case 'workflow_node':
      return 'Workflow node'
    case 'mcp_server':
      return 'MCP server'
  }
}

export function formatLoadDurationMs(durationMs: number): string {
  if (durationMs <= 0) {
    return '—'
  }
  if (durationMs < 1) {
    return '<1 ms'
  }
  if (durationMs < 100) {
    return `${durationMs.toFixed(1)} ms`
  }
  return `${Math.round(durationMs)} ms`
}

export function formatPluginStatus(status: PluginStatus): string {
  return status === 'loaded' ? 'Loaded' : 'Failed'
}
